compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -512,23 +512,18 @@ class SpaceEngine(using Context) extends SpaceLogic {
       if converted == null then tp else ConstantType(converted)
     case _ => tp
 
-  /** Adapt types by performing primitive value boxing.  #12805 */
-  def maybeBox(tp1: Type, tp2: Type): Type =
-    if tp1.classSymbol.isPrimitiveValueClass && !tp2.classSymbol.isPrimitiveValueClass then
-      defn.boxedType(tp1).narrow
-    else tp1
+  def adaptType(tp1: Type, tp2: Type): Type = trace(i"adaptType($tp1, $tp2)", show = true) {
+    def isPrimToBox(tp: Type, pt: Type) =
+      tp.classSymbol.isPrimitiveValueClass && (defn.boxedType(tp).classSymbol eq pt.classSymbol)
+    if      isPrimToBox(tp1, tp2) then defn.boxedType(tp1).narrow   //  1  <:< Integer => (<skolem> : Integer) <:< Integer = true
+    else if isPrimToBox(tp2, tp1) then defn.unboxedType(tp1).narrow // ONE <:< Int     => (<skolem> : Int)     <:< Int     = true
+    else convertConstantType(tp1, tp2)
+  }
 
   /** Is `tp1` a subtype of `tp2`?  */
-  def isSubType(_tp1: Type, tp2: Type): Boolean = {
-    val tp1 = maybeBox(convertConstantType(_tp1, tp2), tp2)
-    //debug.println(TypeComparer.explained(_.isSubType(tp1, tp2)))
-    val res = if (ctx.explicitNulls) {
-      tp1 <:< tp2
-    } else {
-      (tp1 != constantNullType || tp2 == constantNullType) && tp1 <:< tp2
-    }
-    debug.println(i"$tp1 <:< $tp2 = $res")
-    res
+  def isSubType(tp1: Type, tp2: Type): Boolean = trace(i"$tp1 <:< $tp1", debug, show = true) {
+    if tp1 == constantNullType && !ctx.explicitNulls then tp2 == constantNullType
+    else adaptType(tp1, tp2) <:< tp2
```

Comment:
How about:
```suggestion
    else
      isPrimToBox(tp1, tp2) || isPrimToBox(tp2, tp1) ||
      convertConstantType(tp1, tp2) <:< tp2
```
so you don't need to generate skolems just to return true.

Reply:
Yeah, I was starting to think that too...  I'll take your commit and make it compile.  But I also wonder whether it's safer to keep going through `<:<`.

For instance I'm taking the `<:<` out of the constant null branch too, and I'm not sure how wise that is.  And earlier today the patch was failing because it didn't like the difference between `TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Int)` vs `TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int)`...

Reply:
Nope, here's a counter example where this now does the wrong thing:

```
==> Integer <:< (1 : Int)?
<== Integer <:< (1 : Int) = true
```